### ISSUE TITLE
Select lists

### ISSUE THREAD

**Issue by chrisschaub** [signalpoint/DrupalGap#347]
I have a select list that has allowed values. The field draws but the select options are empty. Using the latest 7.x-1.x pull. I do have multiple lists on this form. I've had to hack in a call back to populate them for taxonomy reference terms. But these allowed values are just the default values specified in Drupal's field settings interface.

**Comment by signalpoint** [comment]
@chrisschaub What type of field are you using? What's the widget? Please share the allowed values list.

I'm fairly certain taxonomy term reference fields (even multiple ones) should work.

Hopefully I didn't just break something.

**Comment by chrisschaub** [comment]
I had to do a callback to populate
- Term reference    Select list

and

List (text)    Select list

The allowed values for the last one are ...

AL|Alabama
AK|Alaska
AS|American Samoa
AZ|Arizona
AR|Arkansas
AA|Armed Forces Americas
AP|Armed Forces Pacific
AE|Armed Forces Others
CA|California
CO|Colorado
CT|Connecticut
DE|Delaware
DC|District Of Columbia
FL|Florida
GA|Georgia
GU|Guam
HI|Hawaii
ID|Idaho
IL|Illinois
IN|Indiana
IA|Iowa
KS|Kansas
KY|Kentucky
LA|Louisiana
ME|Maine
MD|Maryland
MA|Massachusetts
MI|Michigan
MN|Minnesota
MS|Mississippi
MO|Missouri
MT|Montana
'NE|Nebraska
NV|Nevada
NH|New Hampshire
NJ|New Jersey
NM|New Mexico
NY|New York
NC|North Carolina
ND|North Dakota
MP|Northern Mariana Islands
OH|Ohio
OK|Oklahoma
OR|Oregon
PA|Pennsylvania
PR|Puerto Rico
RI|Rhode Island
SC|South Carolina
SD|South Dakota
TN|Tennessee
TX|Texas
UM|United States Minor Outlying Islands
UT|Utah
VT|Vermont
VI|Virgin Islands
VA|Virginia
WA|Washington
WV|West Virginia
WI|Wisconsin
WY|Wyoming

Thanks!

On Fri, Aug 22, 2014 at 9:32 AM, Tyler Frankenstein <
notifications@github.com> wrote:

> @chrisschaub https://github.com/chrisschaub What type of field are you
> using? What's the widget? Please share the allowed values list.
> 
> I'm fairly certain taxonomy term reference fields (even multiple ones)
> should work.
> 
> Hopefully I didn't just break something.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/signalpoint/DrupalGap/issues/347#issuecomment-53067450
> .

## 

Christopher Schaub
http://chris.schaub.com

**Comment by signalpoint** [comment]
@chrisschaub Taxonomy term reference fields should be populated automatically by DrupalGap core, you shouldn't have to do a custom callback. I believe DG core may limit it to only 20 options though (the default for the taxonomy term index service resource).

It appears your Nebraska value has a single quote in front of the value, maybe that is breaking it. Try this list instead: https://www.drupal.org/node/332575

I just confirmed on my local host that List (text) / Select list works fine for the U.S. states when it is a single value field.

I also just confirmed that a List (text) / Select list field with 2 or more values, does not work properly on an entity add/edit form.

**Comment by chrisschaub** [comment]
Just used the new list, still no go. Very strange. And I'm using 7.x-1.x latest pull. Is that the best branch?

**Comment by signalpoint** [comment]
@chrisschaub That's the correct branch.

Is your field a single value field?

**Comment by chrisschaub** [comment]
The number of allowed values is set to 1.

On Fri, Aug 22, 2014 at 10:58 AM, Tyler Frankenstein <
notifications@github.com> wrote:

> @chrisschaub https://github.com/chrisschaub That's the correct branch.
> 
> Is your field a single value field?
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/signalpoint/DrupalGap/issues/347#issuecomment-53079149
> .

## 

Christopher Schaub
http://chris.schaub.com

**Comment by signalpoint** [comment]
@chrisschaub I also just confirmed (on localhost) that a content type with a taxonomy term reference field, and a single value List (text) / Select list works just fine.

The taxonomy term reference field is auto populated, as is List (text) field. Creating a node works fine.

Also grab the latest snapshot of the DrupalGap module for your Drupal site, and enable any new sub modules that come with it.

Other than that, I don't know what's going on. Maybe you have some custom form alterations that are interferring (i.e. your callback for the term reference field).

**Comment by chrisschaub** [comment]
In both cases I have more than one select list on the form.

**Comment by signalpoint** [comment]
I just confirmed that multiple taxonomy term reference fields, and multiple List (text) fields work fine on an creation form.

![multiple-terms-and-select-lists](https://cloud.githubusercontent.com/assets/671140/4013556/943ad48a-2a17-11e4-8a7d-cad307159bcb.png)

I also confirmed that I am able to edit that entity just fine. However, the existing values for the taxonomy term reference field aren't auto loaded into the select list, #245, but you can reselect the values, and it'll save just fine. The List (text) are loaded just fine.

**Comment by chrisschaub** [comment]
Great, yes, we had some old code that was borking it. Thanks for confirming this for me. Just curious, I  see "select" as the default option for select lists with allowed values. But taxonomy term reference fields show blank by default. Is that what you see?

**Comment by signalpoint** [comment]
@chrisschaub Are you seeing the blank option for taxonomy term reference fields on a new node, existing node, or both?

**Comment by chrisschaub** [comment]
Just new, haven't tried editing one.

**Comment by chrisschaub** [comment]
I think there is also an issue selecting the first option. You don't see "--select--" like you do for other list fields. So, you have to select twice to select the first option in the list, won't work first time.

**Comment by signalpoint** [comment]
@chrisschaub Yes, I've experienced the "have to select twice" bug before, I figured that may be jQM, not sure though.

As for not seeing "-select", I'm fairly certain that is because the field is marked as required. I tried to implement to field widgets to respond accordingly to the "required" setting in Drupal. I only remember concentrating on select lists, i.e. if it is required, automatically select the first option.

Will you please confirm/describe if the widget responds differently depending on whether or not your field is marked as require?

**Comment by signalpoint** [comment]
@chrisschaub Here is what the Drupal FAPI uses:

```
Required: List (text) / Select list - "_none"|"- Select a value -"
Optional: List (text) / Select list - "_none"|"- None -"
Required: Term reference / Select list - "_none"|"- Select a value -"
Optional: Term reference / Select list - "_none"|"- None -"
```

So they are using `_none` to signal the empty value. I swear they were using an empty string before, oh well. Either way, we need to get the DGFAPI updated for select lists and term reference fields to adhere to this ruling.

In all cases, required or not, Drupal has the _none value and label as the default (when there isn't a default value).

Here's what the DGFAPI currently does for its defaults:

```
Required: List (text) / Select list - ""|"Select"
Optional: List (text) / Select list - ""|"- None -"
Required: Term reference / Select list - null,null (this is the bug you just fixed with #361)
Optional: Term reference / Select list - null,null (this is still a bug)
```

So the DGFAPI needs to be updated to match the Drupal FAPI.

**Comment by signalpoint** [comment]
I just updated the DGFAPI to this:

```
Required: List (text) / Select list - ""|"- Select a value -"
Optional: List (text) / Select list - ""|"- None -"
Required: Term reference / Select list - ""|"- Select a value -"
Optional: Term reference / Select list - ""|"- None -"
```

Now it'd just be a matter of changing the value of the empty string to `_none` around the system, and this would be done. Small fish though, since the empty string works for now.
